### DISCUSSION TITLE
best way to handle delayed fetch of data in a signal store?

### DISCUSSION THREAD

**Discussion by skyleguy** [ngrx/platform#4464] [Q&A] [answered]
hello all.

I have an ngrx store slice which is responsible for holding global info for my app. One field is called `userId` and is set after the user has signed in. I then have a signal store which is focused more on holding the data for one specific feature. when the user id is set in the global store it needs to trigger a different call using that user id to get the information it needs. I have done this like so:

```
withMethods(() => ({
  fetchData(userId: string): ... // my method which uses rxMethod to fetch data and set in signal store
}))
withHooks({
  onInit(store, globalSliceFacade = inject(GlobalSliceFacade)) {
    effect(() => {
      const userId = globalSliceFacade.userId(); // use global slice facade signal selector on my userId field
      if (userId) {
        store.fetchData(userId)
      }
    }, {
      allowSignalWrites: true
    })
  }
})
```

as we can see the signal store uses the onInit hook to set up an effect against the globalSliceFacade's userId signal and if the userId exists we trigger a call to fetch the data. I could not think of another way for this signal store to know when to fetch the data (unless the global facade has an effect on the setUserId action which doesnt dispatch anything and simply calls this signal store, but i didnt want the global store to know about this signal store)

the only reason i dont _love_ this is because it will be the first time that we've had to enable signal writes in our app (we have barely used signals yet). I know that there are valid use cases for enabling it and it is nearly impossible for us to get stuck in an infinite loop here as under no circumstance should this signal store be triggering anything that changes the userId.

So my question is: **is this a valid time to allowSignalWrites or is there a better way to make this signal store wait before fetching its data?**

**Comment by rainerhahnekamp** [comment, marked as answer]
I do it in the same way but with `untracked`. `allowSignalWrites` is going to vanish: https://github.com/angular/angular/pull/56501

It is not a sin to use `untracked` btw. There is complete discussion of making it the default: https://github.com/angular/angular/issues/56155

You could of course create a method with `rxMethod` and call that method via the init hook. That method would at the same time also do the fetching. But I am not sure if it brings that much in terms of readability.

**Reply by skyleguy** [reply]
thanks for the response!

I did not realize that `allowSignalWrites` was being looked at for removal (the fun of using developer preview features). if it becomes the default to allow this then i guess untracked would not even be needed? 

Either way it seems like even if i swap `allowSignalWrites` for `untracked` I may have to redo it soon after once those discussions you linked have been integrated?

**Reply by skyleguy** [reply]
im sure its been discussed and was intentionally left out of the signal store api but a `withEffects` would be kind of nice so we wouldnt need to use the `onInit` hook to set up some effects

**Reply by rainerhahnekamp** [reply]
So I would switch to `untracked` only. Given the answers of the Angular team, I don't expect that explicit tracking will land in the near future.

**Reply by skyleguy** [reply]
thanks for the advice. sounds like a step in the right direction for my implementation 👍

**Reply by rainerhahnekamp** [reply]
Great, btw, if your question has been answered, please mark it.

**Reply by skyleguy** [reply]
was going to wait and see if more discussion opened up but it appears not!
